### DISCUSSION TITLE
Taking the plunge!

### DISCUSSION THREAD

**Discussion by rthorntn** [jglathe/linux_ms_dev_kit#17] [Q&A]
Thanks for this jglathe, absolutely amazing!

So on my thought process...I'd love some feedback.

I'm thinking of using a new NVMe.  Honestly I'd prefer to just have Linux on there, bad idea?

I "might" wait for Pop OS 24.04?

I basically want to use this as my daily Linux driver?

Are there any current gotchas?

Thanks again.
Richard

**Comment by jglathe** [comment]
I would recommend using a bigger SSD, although for a first test round, just plug in an external SSD enclosure via USB-A or type-c and boot from the image. It would be a good idea to de-bitlocker the Windows install, in that case you can fetch the device specific firmwares from the windows installation on the internal nvme (the image does just that on first boot). The external enclosure is usually fast enough for normal use, I've used one box for months this way. 

#### Gotchas

When booting from type-c, orientation of the device matters. For WDK, logo down on external device = normal. In normal orientation you get full speed, in swapped orientation you get only USB2 speeds.

If a Linux boot fails, the thermal managament of the box might switch to emergency mode. The box only cools in full fan blasts when stressed then. To restore the normal smooth behaviour you need to start windows and shut down again.

After going through the install, run `sudo update-grub` and it should discover the Windows install on the internal nvme. It can boot both from the external enclosure.

Pop-OS 24.04 is not ready for prime time, yet, and there is no install ISO image for the WDK (yet). I will update when it's available. Moving about installations between disks can be done with the onboard tools, disks and gparted. Worked fine.

Have fun!

**Comment by jglathe** [comment]
Oh and yes, using it as daily linux driver works fine here. I also have some x64 boxes, but usualy I only use the wdks or X13s.

**Comment by rthorntn** [comment]
Thank you!

Wow, is the USB-C not being wired symmetrically a bug or a feature, maybe they used a USB-A to USB-C converter internally?

Any idea if the thermal management under Linux is something that can be improved?

Thinking out loud...such a low TDP CPU deserves to be fanless, (as long as the system functions with the fan removed) I could ditch the case, mill a plate to attach to the hot bits and attach a suitable heatsink to the plate. 

Update: I updated Winblows, made sure Bitlocker was off, disabled Secureboot and set USB to be the the first boot... it booted Pop OS 22.04 from USB no problems.  AMAZING!

More thinking out loud... on the NVMe front, maybe I should put one of your amazing Linux images on to an NVMe (with a USB to NVMe adapter), USB boot the MDK with it, let it copy all of its firmware and then install the Linux NVMe into the MDK...if I get the fan issue (before I go fanless) I can put the Winblows NVMe back in?

**Reply by jglathe** [reply]
Somebody is using my hacked-up pop! 22.04 image, nice :grin:  Pop-OS is another can of worms, although its GUI is very accomplished. I like it when using it. I actually built the newest 6.10 kernel for pop on the WDK running 22.04 yesterday, will put it up on the LinuxKernels Google Drive.

About those thermals. I have 2 WDKs, one running as daily driver in the living room, one in the home office computer room. I usually don't hear them, except when I'm making build runs, or when the thermal management is in blast mode and I'm doing build runs. In normal operation (even with high load) it's barely audible or completely silent. So, as long as you can boot to Windows once when the blasts happen, it's a non - issue. The background is the EC controller Microsoft is using and some marker at successful boot in the EFI variables that gets set. There are more pressing issues, the thermals of the WDK are fine.

The nvme: Internal nvme is a 2230 model, I'm using a Micron 2400 2TB. The port is PCIe3x4. Writing the image to such a disk and mounting it afterwards will work. If... you also have the Windows install on it, with all of it working via boot manager (grub-efi). Otherwise you will run into the thermal issue headfirst, with lots of work to fix it. Yes you can put back the original NVMe, and it will boot. And rectify the fan issue. But this won't help you with the other disk, as it has its own ESP. 

![Screenshot from 2024-07-25 10-00-10](https://github.com/user-attachments/assets/4acd6157-a3b6-434e-bc9f-490193961975)
This is the partition layout of my X13s, for reference. It is set up with the original Windows installation, Ubuntu 24.04 and Pop!_OS 24.04. I would recommend something similar.

**Reply by rthorntn** [reply]
I read that Pop 24.04 Alpha was out, are all the bits available to do an arm64 image for WDK?

**Reply by jglathe** [reply]
It is out, I have a [repository](https://github.com/jglathe/pop-os-iso/tree/jg/arm64-x13s) that can generate an image for the x13s, and with some further work for the WDK, too. System76 has a branch for the snapdragon x elite :grin: I will get around to doing this eventually. Currently chasing a different bug related to 4k displays on miniDP.

**Reply by jglathe** [reply]
Found it, finally :sunglasses: 4k@60Hz resolution on the wdk looked like this after resuming since ~6.7-ish:
![IMG_8418](https://github.com/user-attachments/assets/4d38b8d6-8ec7-451e-be1c-723718b03549)
I thought this was some bug in power management or graphics driver, but actually it was memory management, specifically CMA region. This is "fixed" now by assigning CMA_SIZE_MBYTES=128 as it should be. What a PITA that was.

**Reply by rthorntn** [reply]
Ripper, well done!

**Reply by jglathe** [reply]
Well it was a little more complicated as it seems. Will need to do a wiki entry on this. The problem is somehow rooted in clock controllers, and appears to be that camcc and venus interfere in that resolution. You usually want venus (video codecs) for performance reasons. For venus to work you also need videocc-sm8350. camcc is on the chip but will not be of any use on the wdk. Only way to have this gracefully appears to be to blacklist camcc_sc8280xp on the wdk. Disabling the unit in the dtb somehow doesn't do. So, with the same kernel on the X13s, it loads camcc and even the camera has a chance on working. That bsod issue also isn't there.

**Comment by rthorntn** [comment]
Thanks, I'll take your advice.

On the NVMe (size) I was thinking of modding the mounting post to allow for a longer drive as I have a 2280 in the spares bin?

**Reply by jglathe** [reply]
I decided against it, but ofc there is no warranty to consider anymore. Since you're fond of metal work, why not. Others have done so. For me buying a drive that fits was the less risky option.

**Comment by jglathe** [comment]
Regarding type-c, swapping and booting from it: That's quite a topic. Of course there is wiring for swapping the orientation. There are also GPIOs to detect the orientation, and GPIOs that switch the multiplexers for the correct wiring and additionally between USB mode(s) and DP altmode. On the QCom SoCs, the functions controlling it are at least partially in the remote processors (Hexagon). They have their own firmware, which needs to be downloaded to them and brought up. 
When you boot from type-c, the EFI BIOS handles the first I/O. And it apparently  doesn't handle the orientation switching, therefore you can't get into EFI BIOS with a type-c connected display only. You will always need a DP connected display. Regarding I/O from the type-c boot device it means that you're kind of stuck. You load kernel, dtb, initramfs, boot the kernel. It discovers the hardware (in whatever order), brings up the nvme, XHCI, remote processors, tries to load firmware (from initramfs), and mounts the rootfs. On all this chain there is no handover from the already configured and used type-c device to re-connect to get orientation switching and proper discovery. It is already in use and "wired". That's why the orientation is important.
There is some work going on to improve on the type-c orientation and role switching. The 6.10 kernel has some patches in it regarding this. But this won't solve the issue of booting from type-c in the right orientation.

**Reply by jglathe** [reply]
In my tree, there is actually code to get orientation info. After plugging the device, you can do `sudo dmesg | tail -n 50 | grep sbu` to get this info.

![Screenshot from 2024-07-25 12-23-31](https://github.com/user-attachments/assets/a0b519c6-d44d-44fc-b991-737fcb69a225)

As I said there is work to improve on this, but boot from type-c will be an orientation issue on the WDK (and those other fancy laptops) for the time being.

**Comment by rthorntn** [comment]
Thanks!

On the 6.10 kernel for Pop OS 22.04 how to install that?

I now have Pop OS 22.04 on an 1TB NVMe booting over USB, could I trouble you for a quick write up (sequence and tools used) for getting Winblows and Pop OS on the same NVMe for internal installation in the SDK, I think Ill set aside 64GB for Winblows and Linux can have the rest. 

Thanks again!

**Reply by jglathe** [reply]
You want to mount the 1TB SSD into the wdk, right? Well then it's the "wrong" order. Let's install new kernel first.
1. Download the [folder](https://drive.google.com/drive/folders/1-ZHZfam-Hs1sxho51fRSU2GlGXe_9A54) with the deb packages. have them all in one separate folter, like `~/Downloads/6.10.0-76061000`. 
2. Open a terminal in this folder, execute `sudo dpkg -i *.deb`.
3. Check in /boot if the symlinks look trustworthy. dtb -> to a 6.10 dtb, initrd.img -> our initrd.img and so on.
4. Take a look at /boot/grub/grub.cfg (need to be sudo to even read). The boot entry for Linux should look similar:

```
menuentry 'Ubuntu' --class ubuntu --class gnu-linux --class gnu --class os $menuentry_id_option 'gnulinux-simple-0ee58b2d-928f-4a38-bf4a-8d7face26345' {
        recordfail
        savedefault
        load_video
        gfxmode $linux_gfx_mode
        insmod gzio
        if [ x$grub_platform = xxen ]; then insmod xzio; insmod lzopio; fi
        insmod part_gpt
        insmod ext2
        search --no-floppy --fs-uuid --set=root 0ee58b2d-928f-4a38-bf4a-8d7face26345
        linux   /boot/vmlinuz-6.10.1-061001-generic root=UUID=0ee58b2d-928f-4a38-bf4a-8d7face26345 ro  rootwait pd_ignore_unused clk_ignore_unused regulator_ignore_unused arm64.nopauth efi=noruntime drm.debug=0x100 loglevel=6 dyndbg='file drivers/base/firmware_loader/main.c +p'
        initrd  /boot/initrd.img-6.10.1-061001-generic
        devicetree      /boot/dtb-6.10.1-061001-generic
}
```
The interesting parts are that all three parts, linux, initrd, and devicetree are loaded. 
If that is the case, reboot and select the new kernel in the grub menu.

**Reply by rthorntn** [reply]
Thanks, by the "wrong" order do you mean I clone Windows onto the new SSD first, resize it and then put Linux on?

**Reply by jglathe** [reply]
Sorta. You usually start with the recovery image on a new SSD. Cloning Windows might work, too, done this, too. But Windows needs to be there first because partition IDs and BCD boot.

**Reply by rthorntn** [reply]
Ah cool thanks, I wasn't aware there was a recovery image for it!

**Reply by jglathe** [reply]
you can obtain it from Microsoft when you're logged in with your MSA (which you are, ofcourse). You need a 32GB stick. With the newest UEFI BIOS it doesn't even re-enable all the secure boot stuff or restore only bitlockered. I used it the other day to track down an issue with nvme and btrfs. There seems to be some kind of pickyness with my WD SN740, but btrfs only.

**Comment by rthorntn** [comment]
I just read https://www.phoronix.com/news/Windows-Dev-Kit-2023-Linux nice work!

So does that mean that any ARM64 distro will work if its installer uses 6.12 or newer?

**Reply by jglathe** [reply]
Oh interesting, haven’t seen it 😀 Yes that would (sort of) be the case. `arch/arm64/config/defconfig` works ootb with the wdk, what you need is boot parameters and (minimal) initramfs module list. There is a lot going on regarding aarch64 boot support with dtb selection, this will be solved. It will be part of the boot chain.
